File: subsystems/elevator.py
"""Elevator subsystem -- 2-stage single-mast elevator (AndyMark "Elevator in
a Box" style cascade rig), spring-assisted extension, motor+rope retraction.

Teaching-bot proof of concept. The Redline motor here is brushed and has no
encoder (a real one could add a through-bore/versa encoder later for
closed-loop positioning -- see README) -- this subsystem is entirely
open-loop, driven by a limit switch at each end of travel. Raising needs
less motor power because the springs are doing most of the work; lowering
needs the motor to actively pull the rope in against that same spring
tension.
"""
from __future__ import annotations

import rev
import wpilib
from commands2 import Command, Subsystem, cmd

from constants import ElevatorConstants


class Elevator(Subsystem):
    def __init__(self) -> None:
        super().__init__()

        self._lift_motor = rev.SparkMax(ElevatorConstants.LIFT_MOTOR_ID, rev.SparkMax.MotorType.kBrushed)
        lift_config = rev.SparkMaxConfig()
        lift_config.inverted(ElevatorConstants.LIFT_MOTOR_INVERTED)
        lift_config.setIdleMode(rev.SparkBaseConfig.IdleMode.kBrake)
        lift_config.smartCurrentLimit(ElevatorConstants.LIFT_CURRENT_LIMIT)
        self._lift_motor.configure(lift_config, rev.ResetMode.kResetSafeParameters, rev.PersistMode.kPersistParameters)

        self._top_limit_switch = wpilib.DigitalInput(ElevatorConstants.TOP_LIMIT_SWITCH_DIO_PORT)
        self._bottom_limit_switch = wpilib.DigitalInput(ElevatorConstants.BOTTOM_LIMIT_SWITCH_DIO_PORT)

    def is_at_top(self) -> bool:
        raw = self._top_limit_switch.get()
        return not raw if ElevatorConstants.TOP_LIMIT_SWITCH_INVERTED else raw

    def is_at_bottom(self) -> bool:
        raw = self._bottom_limit_switch.get()
        return not raw if ElevatorConstants.BOTTOM_LIMIT_SWITCH_INVERTED else raw

    def stop(self) -> None:
        self._lift_motor.set(0.0)

    def raise_command(self) -> Command:
        """Raises while held, stopping automatically at the top even if the
        button is still held (protects the mechanism -- always stop at a
        limit switch, whether the command is "manual" or not)."""

        def _run() -> None:
            if self.is_at_top():
                self.stop()
            else:
                self._lift_motor.set(ElevatorConstants.RAISE_SPEED)

        return cmd.run(_run, self).finallyDo(lambda interrupted: self.stop())

    def lower_command(self) -> Command:
        """Lowers while held, stopping automatically at the bottom."""

        def _run() -> None:
            if self.is_at_bottom():
                self.stop()
            else:
                self._lift_motor.set(ElevatorConstants.LOWER_SPEED)

        return cmd.run(_run, self).finallyDo(lambda interrupted: self.stop())

    def periodic(self) -> None:
        wpilib.SmartDashboard.putBoolean("Elevator/AtTop", self.is_at_top())
        wpilib.SmartDashboard.putBoolean("Elevator/AtBottom", self.is_at_bottom())

File: src/main/java/frc/robot/subsystems/Elevator.java
package frc.robot.subsystems;

import static frc.robot.Constants.ElevatorConstants.*;

import com.revrobotics.PersistMode;
import com.revrobotics.ResetMode;
import com.revrobotics.spark.SparkLowLevel.MotorType;
import com.revrobotics.spark.SparkMax;
import com.revrobotics.spark.config.SparkBaseConfig.IdleMode;
import com.revrobotics.spark.config.SparkMaxConfig;

import edu.wpi.first.wpilibj.DigitalInput;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.SubsystemBase;

/**
 * Elevator subsystem -- 2-stage single-mast elevator (AndyMark "Elevator in a Box"
 * style cascade rig), spring-assisted extension, motor+rope retraction.
 *
 * <p>Teaching-bot proof of concept. The Redline motor here is brushed and has no
 * encoder (a real one could add a through-bore/versa encoder later for closed-loop
 * positioning -- see README) -- this subsystem is entirely open-loop, driven only by a
 * limit switch at each end of travel. Raising needs less motor power because the
 * springs are doing most of the work; lowering needs the motor to actively pull the
 * rope in against that same spring tension.
 *
 * <p>This subsystem only exposes plain actions (raise/lower a notch, stop, check the
 * limit switches) -- the "keep raising/lowering while a button is held, but always stop
 * at a limit switch even if the button is still held" behavior lives in
 * commands/RaiseElevatorCommand.java and commands/LowerElevatorCommand.java.
 */
public class Elevator extends SubsystemBase {

    // kBrushed, not kBrushless: a Redline motor has physical brushes (hence the name)
    // and no built-in encoder, unlike every NEO in this project. SparkMax can drive
    // either motor type, but has to be told which one it's talking to, since brushed
    // and brushless motors are commutated (have their windings energized in sequence)
    // completely differently in hardware.
    private final SparkMax liftMotor = new SparkMax(LIFT_MOTOR_ID, MotorType.kBrushed);

    private final DigitalInput topLimitSwitch = new DigitalInput(TOP_LIMIT_SWITCH_DIO_PORT);
    private final DigitalInput bottomLimitSwitch = new DigitalInput(BOTTOM_LIMIT_SWITCH_DIO_PORT);

    public Elevator() {
        SparkMaxConfig liftConfig = new SparkMaxConfig();
        liftConfig.inverted(LIFT_MOTOR_INVERTED);
        liftConfig.idleMode(IdleMode.kBrake);
        liftConfig.smartCurrentLimit(LIFT_CURRENT_LIMIT);
        liftMotor.configure(liftConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);
    }

    public boolean isAtTop() {
        boolean raw = topLimitSwitch.get();
        return TOP_LIMIT_SWITCH_INVERTED ? !raw : raw;
    }

    public boolean isAtBottom() {
        boolean raw = bottomLimitSwitch.get();
        return BOTTOM_LIMIT_SWITCH_INVERTED ? !raw : raw;
    }

    /** speed is a duty cycle in [-1, 1]: positive raises, negative lowers -- see
     * {@code ElevatorConstants.RAISE_SPEED}/{@code LOWER_SPEED}. */
    public void setSpeed(double speed) {
        liftMotor.set(speed);
    }

    public void stop() {
        liftMotor.set(0.0);
    }

    @Override
    public void periodic() {
        SmartDashboard.putBoolean("Elevator/AtTop", isAtTop());
        SmartDashboard.putBoolean("Elevator/AtBottom", isAtBottom());
    }
}
